File: app/src/main/java/com/pij/zworkout/workout/viewmodel/HorrocksWorkoutViewModel.java
package com.pij.zworkout.workout.viewmodel;

import android.support.annotation.NonNull;

import com.annimon.stream.Optional;
import com.pij.horrocks.AsyncInteraction;
import com.pij.horrocks.Configuration;
import com.pij.horrocks.Engine;
import com.pij.horrocks.Interaction;
import com.pij.horrocks.MultipleReducerCreator;
import com.pij.horrocks.ReducerCreator;
import com.pij.horrocks.SingleReducerCreator;
import com.pij.horrocks.Storage;
import com.pij.utils.Logger;
import com.pij.zworkout.service.api.WorkoutFile;
import com.pij.zworkout.uc.Workout;
import com.pij.zworkout.workout.Model;
import com.pij.zworkout.workout.State;
import com.pij.zworkout.workout.WorkoutViewModel;

import org.jetbrains.annotations.NotNull;

import io.reactivex.Observable;

import static java.util.Arrays.asList;

/**
 * <p>Created on 01/03/2018.</p>
 *
 * @author Pierrejean
 */

public class HorrocksWorkoutViewModel implements WorkoutViewModel {

    private final Logger logger;
    private final Observable<Model> modelStream;
    private final ReducerCreator<String, State> loader;
    private final ReducerCreator<Object, State> createWorkout;
    private final ReducerCreator<Object, State> save;
    private final ReducerCreator<String, State> name;

    private HorrocksWorkoutViewModel(Logger logger,
                                     Engine<State, Model> engine,
                                     Storage<State> storage,
                                     ReducerCreator<String, State> nameFeature,
                                     ReducerCreator<String, State> loadingFeature,
                                     ReducerCreator<Object, State> saveFeature,
                                     ReducerCreator<Object, State> createWorkoutFeature
    ) {
        this.logger = logger;

        name = nameFeature;
        loader = loadingFeature;
        save = saveFeature;
        createWorkout = createWorkoutFeature;
        Configuration<State, Model> engineConfiguration = Configuration.<State, Model>builder()
                .store(storage)
                .transientResetter(this::resetTransient)
                .stateToModel(this::convert)
                .creators(asList(name, loader, save, createWorkout))
                .build();
        modelStream = engine.runWith(engineConfiguration).share();
    }

    /**
     * Helper constructor.
     */
    public static HorrocksWorkoutViewModel create(Logger logger,
                                                  Engine<State, Model> engine,
                                                  Storage<State> storage,
                                                  Interaction<String, State> nameFeature,
                                                  AsyncInteraction<String, State> loadingFeature,
                                                  AsyncInteraction<Object, State> saveFeature,
                                                  Interaction<Object, State> createWorkoutFeature
    ) {
        return new HorrocksWorkoutViewModel(logger, engine, storage,
                new SingleReducerCreator<>(nameFeature, logger),
                new MultipleReducerCreator<>(loadingFeature, logger),
                new MultipleReducerCreator<>(saveFeature, logger),
                new SingleReducerCreator<>(createWorkoutFeature, logger)
        );
    }

    public static State initialState() {
        return State.builder()
                .inProgress(false)
                .showError(Optional.empty())
                .showSaved(false)
                .workout(Workout.EMPTY)
                .file(WorkoutFile.UNDEFINED)
                .build();
    }

    @NonNull
    private State resetTransient(@NonNull State input) {
        return input.toBuilder()
                // TODO code that
                .build();
    }

    @NonNull
    private Model convert(@NonNull State state) {
        return Model.builder()
                .inProgress(state.inProgress())
                .showSaved(state.showSaved())
                .showError(state.showError())
                .name(state.workout().name())
                .build();
    }

    @Override
    public void load(@NonNull String itemId) {
        loader.trigger(itemId);
    }

    @Override
    public void createWorkout() {
        createWorkout.trigger(new Object());
    }

    @NotNull
    @Override
    public Observable<Model> model() {
        return modelStream
                .doOnError(e -> logger.print(getClass(), "Terminal Damage!!!", e))
                .doOnComplete(() -> logger.print(getClass(), "model() completed!!!"))
                ;
    }

    @Override
    public void name(@NotNull String newValue) {
        name.trigger(newValue);
    }

    @Override
    public void save() {
        save.trigger(new Object());
    }
}

File: app/src/main/java/com/pij/zworkout/workout/HorrocksWorkoutViewModel.kt
/*
 * Copyright (c) 2018, Chiswick Forest
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and limitations under the License.
 */

package com.pij.zworkout.workout

import com.pij.horrocks.*
import com.pij.utils.Logger
import com.pij.zworkout.uc.Workout
import com.pij.zworkout.workout.feature.InsertEffortFeature
import io.reactivex.Observable

/**
 *
 * Created on 01/03/2018.
 *
 * @author Pierrejean
 */

internal class HorrocksWorkoutViewModel private constructor(private val logger: Logger,
                                                            engine: Engine<State, Model>,
                                                            storage: Storage<State>,
                                                            private val converter: StateConverter<State, Model>,
                                                            private val name: ReducerCreator<String, State>,
                                                            private val description: ReducerCreator<String, State>,
                                                            private val loader: ReducerCreator<String, State>,
                                                            private val save: ReducerCreator<Any, State>,
                                                            private val insertEffort: ReducerCreator<Pair<ModelEffort, Int>, State>,
                                                            private val setEffort: ReducerCreator<Pair<ModelEffort, Int>, State>,
                                                            private val editEffortProperty: ReducerCreator<EffortPropertyEvent, State>,
                                                            private val createWorkout: ReducerCreator<Any, State>
) : WorkoutViewModel {
    private val models: Observable<Model>

    init {
        val engineConfiguration = Configuration.builder<State, Model>()
                .store(storage)
                .transientResetter { resetTransient(it) }
                .stateToModel { converter.convert(it) }
                .creators(listOf(name, description, loader, save, insertEffort, setEffort, editEffortProperty, createWorkout))
                .build()
        models = engine.runWith(engineConfiguration).share()
    }

    private fun resetTransient(input: State): State {
        return input.copy(
                showError = null,
                showSaved = false
                // TODO code that
        )
    }

    override fun load(itemId: String) {
        loader.trigger(itemId)
    }

    override fun createWorkout() {
        createWorkout.trigger(Any())
    }

    override fun model(): Observable<Model> {
        return models
                .doOnError { e -> logger.print(javaClass, e, "Terminal Damage!!!") }
                .doOnComplete { logger.print(javaClass, "model() completed!!!") }
    }

    override fun name(newValue: String) {
        name.trigger(newValue)
    }

    override fun description(newValue: String) {
        description.trigger(newValue)
    }

    override fun save() {
        save.trigger(Any())
    }

    override fun addEffort() {
        insertEffort.trigger(Pair(ModelSteadyState(120, "Z1"), InsertEffortFeature.END_OF_LIST))
    }

    override fun setEffort(effort: ModelEffort, position: Int) {
        setEffort.trigger(Pair(effort, position))
    }

    override fun editEffortProperty(description: EffortPropertyEvent) {
        editEffortProperty.trigger(description)
    }

    override fun changeEffortProperty(description: EffortPropertyEvent) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    companion object {

        /**
         * Helper constructor.
         */
        fun create(logger: Logger,
                   engine: Engine<State, Model>,
                   storage: Storage<State>,
                   converter: StateConverter<State, Model>,
                   nameFeature: Interaction<String, State>,
                   descriptionFeature: Interaction<String, State>,
                   loadingFeature: AsyncInteraction<String, State>,
                   saveFeature: AsyncInteraction<Any, State>,
                   insertEffortFeature: Interaction<Pair<ModelEffort, Int>, State>,
                   setEffortFeature: Interaction<Pair<ModelEffort, Int>, State>,
                   editEffortPropertyFeature: Interaction<EffortPropertyEvent, State>,
                   createWorkoutFeature: Interaction<Any, State>
        ): HorrocksWorkoutViewModel {
            return HorrocksWorkoutViewModel(logger, engine, storage,
                    converter,
                    SingleReducerCreator(nameFeature, logger),
                    SingleReducerCreator(descriptionFeature, logger),
                    MultipleReducerCreator(loadingFeature, logger),
                    MultipleReducerCreator(saveFeature, logger),
                    SingleReducerCreator(insertEffortFeature, logger),
                    SingleReducerCreator(setEffortFeature, logger),
                    SingleReducerCreator(editEffortPropertyFeature, logger),
                    SingleReducerCreator(createWorkoutFeature, logger)
            )
        }

        fun initialState(): State {
            return State(inProgress = false,
                    showError = null,
                    showSaved = false,
                    workout = Workout(),
                    file = null,
                    nameIsReadOnly = false)
        }
    }
}
